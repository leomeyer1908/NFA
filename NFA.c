#include "NFA.h"

void addTransitionNFANode(NFANode* nfaNode, char transitionChar, NFANode* destinationNode) {
	LinkedList* possibleDestinations;
	if (containsHashMap(nfaNode->transitions, (int) transitionChar)) {
		possibleDestinations = (LinkedList*) getHashMap(nfaNode->transitions, (int) transitionChar);
	}
	else {
		possibleDestinations = (LinkedList*) malloc(sizeof(LinkedList));
		initList(possibleDestinations);
		insertHashMap(nfaNode->transitions, (int) transitionChar, (void*) possibleDestinations);
	}
	pushBackList(possibleDestinations, (void*) destinationNode);
}

int getIndexNFANode(NFA* nfa, NFANode* nfaNode) {
    for (int i = 0; i < nfa->nodeNum; i++) {
        if (nfa->NFANodes[i] == nfaNode) {
            return i;
        } 
    }
    return -1;
}

void printNFA(NFA* nfa) {
	for (int i = 0; i < nfa->nodeNum; i++) {
        printf("Node Num: %d:\n", i);
        printf("Is Accept: %d\n", nfa->NFANodes[i]->is_accept);
		for (int j = 0; j < nfa->NFANodes[i]->transitions->capacity; j++) {
			for (Node* node = nfa->NFANodes[i]->transitions->array[j].head; node != NULL; node = node->next) {
				char currentTransitionChar = (char) ((KeyValuePair*) node->value)->key;
                LinkedList* transitionList = (LinkedList*) ((KeyValuePair*) node->value)->value;
                if (transitionList != NULL) {
                    for (Node* node = transitionList->head; node != NULL; node = node->next) {
                        NFANode* nextNode = (NFANode*) node->value;
                        printf("'%c' transitions to node #%d\n", currentTransitionChar, getIndexNFANode(nfa, nextNode));
                    }
                }
			}
		}
        printf("\n");
	}
}

void destroyNFANode(NFANode* nfaNode) {
	for (size_t i = 0; i < nfaNode->transitions->capacity; i++) {
		for (Node* node = nfaNode->transitions->array[i].head; node != NULL; node = node->next) {
			LinkedList* transitionList = (LinkedList*) ((KeyValuePair*) node->value)->value;
			if (transitionList != NULL) {
				free(transitionList);
			}
		}
	}
	destroyHashMap(nfaNode->transitions);
	free(nfaNode->transitions);
}

void destroyNFA(NFA* nfa) {
	for (size_t i = 0; i < nfa->nodeNum; i++) {
		destroyNFANode(nfa->NFANodes[i]);
		free(nfa->NFANodes[i]);
	}
}