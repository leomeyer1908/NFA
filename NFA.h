#ifndef NFA_H
#define NFA_H

#include <stdlib.h>
#include <stdio.h>
#include "DataStructures/Stack.h"
#include "DataStructures/HashMap.h"
#include "DataStructures/LinkedList.h"
#include "DataStructures/HashSet.h"

#define EPSILON '\0'

typedef struct NFANode {
    int is_accept;
    HashMap* transitions;
} NFANode;

typedef struct NFA {
    NFANode* startState;
	size_t nodeNum;
	NFANode** NFANodes;
} NFA;


/*
Takes a RegEx with the operators '+', '|', and '*' in the first argument and creates an NFA, which
is stored in the second argument. The NFA is made by first creating an intermediate expression tree 
from the regex rexpression. Then, Thompson's construction is used to build up an NFA by traversing 
thorugh the tree in post order. The function returns 0 for succesful output, and -1 for error.
*/
int createNFAFromRegex(const char* regex, NFA* nfa);

/*Detect if the Regex has any syntax errors. Returns 0 if it has no errors and -1 if it does.*/
int detectErrorRegEx(const char* regex);

void eliminateEpisilonNFA(NFA* nfa);

void addTransitionNFANode(NFANode* nfaNode, char transitionChar, NFANode* destinationNode);

/*Uses a DFS traversal to find all reachable nodes and remove the ones not found*/
void removeUnreachableNFANodes(NFA* nfa);

void printNFA(NFA* nfa);

int getIndexNFANode(NFA* nfa, NFANode* nfaNode);

void destroyNFANode(NFANode* nfaNode);

void destroyNFA(NFA* nfa);

#endif //NFA_H