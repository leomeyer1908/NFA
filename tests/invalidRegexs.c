#include "../NFA.h"

int main() {
    char* arrayOfInvalidRegexs[] = {
        "",
        "a+((a + b)|d",
        "(a+a",
        "()+a",
        "(|a",
        "(+a",
        "(*+a",
        "(",
        ")",
        "(a)(a)",
        "(a)\\+",
        "(a)a",
        "+a",
        "(a+)",
        "a++a",
        "a+|a",
        "a+*",
        "a+",
        "|a",
        "(a|)",
        "a|+a",
        "a||a",
        "a|*",
        "a|",
        "*+a",
        "a*(a)",
        "a*\\+",
        "a*a",
        "\\a",
        "\\ . ",
        "\\+(a)",
        "\\|\\(",
        "\\)a",
        "a(a)",
        "a\\*",
        "ab",
        NULL
    };
    for (int i = 0; arrayOfInvalidRegexs[i] != NULL; i++) {
        printf("Current RegEx: \"%s\",", arrayOfInvalidRegexs[i]);
        NFA nfa;
        int errorResult = createNFAFromRegex(arrayOfInvalidRegexs[i], &nfa);
        if (errorResult == -1) {
            printf(" result: Invalid RegEx.\n");
        }
        else {
            printf(" result: Valid RegEx (Should not happen).\n");
        }
    }
    return 0;
}