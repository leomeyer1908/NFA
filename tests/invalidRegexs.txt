empty
a+((a + b)|d
(a+a
()+a
(|a
(+a
(*+a
(
)
(a)(a)
(a)\\+
(a)a
+a
(a+)
a++a
a+|a
a+*
a+
|a
(a|)
a|+a
a||a
a|*
a|
*+a
a*(a)
a*\\+
a*a
\\a
\\ . 
\\+(a)
\\|\\(
\\)a
a(a)
a\\*
ab