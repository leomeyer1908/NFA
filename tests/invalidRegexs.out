Current RegEx: "", result: Invalid RegEx.
Current RegEx: "a+((a + b)|d", result: Invalid RegEx.
Current RegEx: "(a+a", result: Invalid RegEx.
Current RegEx: "()+a", result: Invalid RegEx.
Current RegEx: "(|a", result: Invalid RegEx.
Current RegEx: "(+a", result: Invalid RegEx.
Current RegEx: "(*+a", result: Invalid RegEx.
Current RegEx: "(", result: Invalid RegEx.
Current RegEx: ")", result: Invalid RegEx.
Current RegEx: "(a)(a)", result: Invalid RegEx.
Current RegEx: "(a)\+", result: Invalid RegEx.
Current RegEx: "(a)a", result: Invalid RegEx.
Current RegEx: "+a", result: Invalid RegEx.
Current RegEx: "(a+)", result: Invalid RegEx.
Current RegEx: "a++a", result: Invalid RegEx.
Current RegEx: "a+|a", result: Invalid RegEx.
Current RegEx: "a+*", result: Invalid RegEx.
Current RegEx: "a+", result: Invalid RegEx.
Current RegEx: "|a", result: Invalid RegEx.
Current RegEx: "(a|)", result: Invalid RegEx.
Current RegEx: "a|+a", result: Invalid RegEx.
Current RegEx: "a||a", result: Invalid RegEx.
Current RegEx: "a|*", result: Invalid RegEx.
Current RegEx: "a|", result: Invalid RegEx.
Current RegEx: "*+a", result: Invalid RegEx.
Current RegEx: "a*(a)", result: Invalid RegEx.
Current RegEx: "a*\+", result: Invalid RegEx.
Current RegEx: "a*a", result: Invalid RegEx.
Current RegEx: "\a", result: Invalid RegEx.
Current RegEx: "\ . ", result: Invalid RegEx.
Current RegEx: "\+(a)", result: Invalid RegEx.
Current RegEx: "\|\(", result: Invalid RegEx.
Current RegEx: "\)a", result: Invalid RegEx.
Current RegEx: "a(a)", result: Invalid RegEx.
Current RegEx: "a\*", result: Invalid RegEx.
Current RegEx: "ab", result: Invalid RegEx.
